core/src/com/mygdx/pirategame/Tornado.java

```diff
@@ -0,0 +1,156 @@
+ * @version 1.0
+ */
+public class Tornado extends Entity {
+
+    private float state = 0;
+    private Animation swirl;
+    private Texture tornado;
+    private Sound windSound;
+    public SpriteBatch batch;
+    private Random rand = new Random();
+    protected boolean inContact;
+    private float damage = 1, timeElapsed = 0f;
+
+    /**
+     * Instantiates a new Tornado.
+     *
+     * @param screen the screen its going onto
+     * @param x      the x value to be placed at
+     * @param y      the y value to be placed at
+     */
+    public Tornado(GameScreen screen, float x, float y) {
+        super(screen, x, y);
+        // Set posistion and size of the tornado
+        int dimension = rand.nextInt(257-128)+128;
+        // The float number used for the height is to keep the ration of the texture image intact
+        setBounds(0,0, dimension / PirateGame.PPM, dimension * 1.473333333f / PirateGame.PPM);
+        //Sets origin of the tornado
+        setOrigin(24 / PirateGame.PPM,24 / PirateGame.PPM);
+        tornado =new Texture(Gdx.files.internal("TornadoSwirls.png"));
+        TextureRegion[][] tmp = new TextureRegion(tornado).split(150, 221);
+        setRegion(tmp[0][0]);
+        swirl = new Animation(0.25f, tmp[0]);
+    }
+
+    /**
+     * Defines an entity
+     */
+    @Override
+    protected void defineEntity () {
+        //sets the body definitions
+        BodyDef bdef = new BodyDef();
+        bdef.position.set(getX(), getY());
+        bdef.type = BodyDef.BodyType.StaticBody;
+        b2body = world.createBody(bdef);
+
+        // Defines a player's shape and contact borders
+        FixtureDef fdef = new FixtureDef();
+        CircleShape shape = new CircleShape();
+        shape.setRadius(150 / PirateGame.PPM);
+
+        // setting BIT identifier
+        fdef.filter.categoryBits = PirateGame.TORNADO_BIT;
+
+        // determining what this BIT can collide with
+        fdef.filter.maskBits = PirateGame.DEFAULT_BIT | PirateGame.PLAYER_BIT | PirateGame.ENEMY_BIT;
+        fdef.shape = shape;
+        fdef.isSensor = true;
+        b2body.createFixture(fdef).setUserData(this);
+    }
+
+    /**
+     * Defines contact
+     */
+    @Override
+    public void entityContact () {
+        inContact = true;
+    }
+
+    /**
+     * Draw the tornados on the gamescreen
+     */
+
+    public void draw () {
+        super.draw(batch);
+    }
+
+    /**
+     * Updates the tornado's position and animation
+     * @param dt time elapsed
+     */
+
+    public void update(float dt){
+        setPosition(b2body.getPosition().x - getWidth() / 2f, b2body.getPosition().y - getHeight() / 2f);
+        setRegion(getFrame(dt));
+        tornadoDamage(dt);
+    }
+
+    /**
+     * Creates a vector force applied on the player if it's close to a tornado. This method also makes the boat turn
+     * @param player The player's instance
+     */
+
+    public void tornadoImpulse(Player player, float dt){
```

Comment:
Nice - good way of working this applying impulse on every GS update...

We just have to make sure of two things (which will be fine really)...

1. That the GS updates are called at regular intervals, otherwise weird background logic calling GS updates would affect the overall impulse on ships, making this unreliable (we could call it 'realistic wind effect' haha 👀 )
2. Maybe implement a distance cutoff for impulse... if a ship is miles away, should it be impacted (physics says yes, but just such a little amount that it's not noticeable... but, what's the point in us using processing time on this?)

Unless I'm missing something?? 😄

Reply:
Yes, you're right in both your points. This is the result of me just trying things and see if they worked tbh, but that means that we can improve it quite a lot. About the first point you said, I understand what you mean, would you know how to check that is what's happening? Could we create a JUnit test for that?